### ISSUE TITLE
Clarification on how to set hyperparameters required with examples.

### ISSUE THREAD

**Issue by erotavlas** [explosion/spaCy#3352]
The documentation is unclear how to set hyperparameters and how spacy uses them when they are set as environment variables.  It also includes an example with some command line arguments that are not present anywhere in the documentation thus their meaning is not known.  What is parser_hidden_depth? What is parser_maxout_pieces?

For example
`parser_hidden_depth=2 parser_maxout_pieces=1 spacy train [...]`

From page
https://spacy.io/api/cli#train-hyperparams

Also it would be nice if some mention about hyperparameter tuning was made and how to achieve this with the current API.

**Comment by theudas** [comment]
I had the same issue with the text classifier and had to read through the source to understand how the hyperparameters work.  
I believe these implementations  are constantly changing. It would still be nice if there was an explanation of the configuration possibilities.

**Comment by honnibal** [comment]
Sorry for the confusion on this. That example should be updated, because neither of those hyper-parameters are good examples of things you'll want to experiment with, usually.

The `parser_hidden_depth` hyper-parameter is currently non-functional. I added that when I was experimenting with the architecture, but ended up hard-coding the depth so that part of the inference could be implemented in Cython, for efficiency.

The `parser_maxount_pieces` parameter controls a dimension of the maxout layer used as part of the parser model. It's best left at 2 or 3 --- it doesn't have that big an impact on accuracy, and higher settings will usually just make the model a bit slower.

I agree that it would be good to get these things cleaned up. The architecture is a bit more stable now, so we could do a bit more to make this easy. This is especially true for the hyper-parameters that are most important to tune, such as the learning rate, optimizer settings, batch size, and dropout.

**Comment by samhardyhey** [comment]
To this end, is hyperparameter optimization on the roadmap at all? I've thought about slapping together something with hyperopt.

**Comment by abmitra84** [comment]
I am also having a bit of struggle to find out how to pass hyperparameters. Consider two scenarios:

Notebook: I want to start `learn_rate` at 0.001 and pull down to 0.0001. I could not send a different learning rate for each epoch. `os.envrion["learn_rate"] = learning_rate` does not work as `os.environ()` requires a string to be passed, if we pass a `str(learning_rate)` instead the script fails

CLI: In CLI set up too I see we can send only single learning rate. Not sure how to set up for different rates with progress of the epoch

@honnibal  @ines , kindly help ..

**Comment by honnibal** [comment]
@abmitra84 You should be able to modify the learning rate explicitly, by writing to the `optimizer.alpha` attribute that you pass into the `nlp.update()` calls. This allows you to change the learning rate dynamically.

**Comment by afkoeppel** [comment]
I've been looking into the possibility of using hyperband to tune the parameters for training a new NER category on a custom training dataset.  However, I'm a bit inexperienced at training and tuning models in general and could benefit some clarification about what the best approach is,  and what parameters can or should be tuned.  For example, right now my code for training the model uses the compounding function for batch size, and the decaying function for drop-out as recommended. I'm not sure if I should be tuning the maximum batch size and/or the initial dropout rate?  Some early exploration suggests the initial drop-out rate isn't having much effect, but max batch size is (I'm getting notably better F1 scores for the new category with max batch 16 rather than max batch 8).   What I'd like to do is use hyperband (or similar) to intelligently explore the parameter space rather than the sort of brute force tuning that I've been doing.  Any advice or recommendations on how best to do this when adding a category to a spaCy NER model would be very helpful.

**Comment by honnibal** [comment]
@afkoeppel The big problem for automated hyper-parameter tuning is trying to express the parameters in a way that they're more independent, so that the cost surface is more convex. This is kind of an under-discussed problem, as the literature usually presents the parameters in the most intuitive way, rather than the way that is best for tuning.

I don't have a good solution for this, but you might want to try having some formula for relating the batch size, learning rate and dropout, and then change the parameters of that formula instead. For instance, some people found that multiplying the learning rate by the square root of the batch size helps. If you change the Adam parameters, you'll probably want to do it by ratio of b1 to b2, rather than letting them vary independently.

Again, this gets sort of out-of-scope for spaCy, though. I don't think anyone has a great solution for it.
